cmd/containerd-shim-runhcs-v1/rootfs.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/layers"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/mount"
+)
+
+// validateRootfsAndLayers checks to ensure we have appropriate information
+// for setting up the container's root filesystem. It ensures the following:
+// - One and only one of Rootfs or LayerFolders can be provided.
+// - If LayerFolders are provided, there are at least two entries.
+// - If Rootfs is provided, there is a single entry and it does not have a Target set.
+func validateRootfsAndLayers(rootfs []*types.Mount, layerFolders []string) error {
+	if len(rootfs) > 0 && len(layerFolders) > 0 {
+		return fmt.Errorf("cannot pass both a rootfs mount and Windows.LayerFolders: %w", errdefs.ErrFailedPrecondition)
+	}
+	if len(rootfs) == 0 && len(layerFolders) == 0 {
+		return fmt.Errorf("must pass either a rootfs mount or Windows.LayerFolders: %w", errdefs.ErrFailedPrecondition)
+	}
+	if len(layerFolders) > 0 && len(layerFolders) < 2 {
+		return fmt.Errorf("must pass at least two Windows.LayerFolders: %w", errdefs.ErrFailedPrecondition)
+	}
+	if len(rootfs) > 1 || rootfs[0].Target != "" {
+		return fmt.Errorf("expected a single rootfs mount with no Target path: %w", errdefs.ErrFailedPrecondition)
+	}
+	return nil
+}
+
+// parseLegacyRootfsMount parses the rootfs mount format that we have traditionally
+// used for both Linux and Windows containers.
+// The mount format consists of:
+//   - The scratch folder path in m.Source, which contains sandbox.vhdx.
+//   - A mount option in the form parentLayerPaths=<JSON>, where JSON is an array of
+//     string paths to read-only layer directories. The exact contents of these layer
+//     directories are intepreteted differently for Linux and Windows containers.
+func parseLegacyRootfsMount(m *types.Mount) (string, []string, error) {
+	// parentLayerPaths are passed in layerN, layerN-1, ..., layer 0
+	//
+	// The OCI spec expects:
+	//   layerN, layerN-1, ..., layer0, scratch
+	var parentLayerPaths []string
+	for _, option := range m.Options {
+		if strings.HasPrefix(option, mount.ParentLayerPathsFlag) {
+			err := json.Unmarshal([]byte(option[len(mount.ParentLayerPathsFlag):]), &parentLayerPaths)
+			if err != nil {
+				return "", nil, fmt.Errorf("unmarshal parent layer paths from mount: %w: %w", err, errdefs.ErrFailedPrecondition)
+			}
```

Comment:
`break` after this or do you still need to parse through remaining options?

Reply:
I've changed to break here, since it is legacy parsing. For the new layer type we will error if passed an unknown option.